synapse/federation/transport/server/__init__.py

```diff
@@ -283,7 +285,48 @@ async def on_GET(
                 },
             )
 
-        return 200, {"sub": user_id}
+        user_id, userinfo_fields = token_info
+        userinfo: JsonDict = {"sub": user_id}
+
+        if userinfo_fields:
+            if OpenIdUserInfoField.DISPLAY_NAME in userinfo_fields:
+                localpart = UserID.from_string(user_id).localpart
+                userinfo[
+                    OpenIdUserInfoField.DISPLAY_NAME
+                ] = await self.hs.get_datastore().get_profile_displayname(localpart)
+            if OpenIdUserInfoField.AVATAR_URL in userinfo_fields:
+                localpart = UserID.from_string(user_id).localpart
+                userinfo[
+                    OpenIdUserInfoField.AVATAR_URL
+                ] = await self.hs.get_datastore().get_profile_avatar_url(localpart)
+            if OpenIdUserInfoField.ROOM_POWERLEVELS in userinfo_fields:
+                userinfo[
+                    OpenIdUserInfoField.ROOM_POWERLEVELS
+                ] = await self._get_powerlevels(user_id)
+
+        return 200, userinfo
+
+    async def _get_powerlevels(self, user_id: str) -> Dict[str, Dict[str, Any]]:
+        room_list = (
+            await self.hs.get_datastore().get_rooms_for_local_user_where_membership_is(
+                user_id=user_id, membership_list=[Membership.JOIN]
+            )
+        )
+
+        power_levels = {}
+        for room in room_list:
+            room_state = await self.hs.get_state_handler().get_current_state(
+                room.room_id
+            )
+            room_power_levels = room_state.get((EventTypes.PowerLevels, ""))
+            if room_power_levels is not None:
+                stripped_content = copy.deepcopy(room_power_levels.content)
+                stripped_content["users"] = {
+                    k: v for k, v in stripped_content["users"].items() if k == user_id
+                }
+                power_levels[room.room_id] = stripped_content
```

Comment:
At this point i followed @turt2live's suggestion to include the whole state event in verbatim, and to include only the current user's powerlevels in the `users` field. See https://github.com/matrix-org/matrix-spec-proposals/pull/3356#discussion_r755574795

Would it meet your concerns if we just froze the currently copied fields/hierarchy? I.e.:
```json
{
    "users": {
      "@jondoe:example.com": 100
    },
    "users_default": 0,
    "events": {
      "m.room.name": 50,
      "m.room.power_levels": 100,
      "m.room.history_visibility": 100,
      "m.room.canonical_alias": 50,
      "m.room.avatar": 50,
      "m.room.tombstone": 100,
      "m.room.server_acl": 100,
      "m.room.encryption": 100
    },
    "events_default": 0,
    "state_default": 50,
    "ban": 50,
    "kick": 50,
    "redact": 50,
    "invite": 0,
    "historical": 100
}
```

Reply:
I don't think I have strong feelings on this. I think the current behaviour is probably fine. It would be nice if the MSC made the behaviour here explicit though.